contracts/Challenge.sol

```diff
@@ -108,107 +110,175 @@ contract Challenge {
     if (blockNumberNHash == bytes32(0) || blockNumberNp1Hash == bytes32(0)) {
       revert("block number too old to challenge");
     }
-    require(blockNumberNp1Hash == computedBlockHash, "end block hash wrong");
+    require(blockNumberNp1Hash == computedBlockHash, "incorrect header supplied for block N+1");
 
-    // decode the blocks
+    // Decode the N+1 block header to construct the fault proof program's input hash.
+    // Because the input hash is constructed from data proven against on-chain block hashes,
+    // it is provably correct, and we can consider that both parties agree on it.
     bytes32 inputHash;
     {
-      Lib_RLPReader.RLPItem[] memory blockNp1 = Lib_RLPReader.readList(blockHeaderNp1);
-      bytes32 parentHash = Lib_RLPReader.readBytes32(blockNp1[0]);
+      Lib_RLPReader.RLPItem[] memory decodedHeader = Lib_RLPReader.readList(blockHeaderNp1);
+
+      bytes32 parentHash = Lib_RLPReader.readBytes32(decodedHeader[0]);
+      // This should never happen, as we validated the hashes beforehand.
       require(blockNumberNHash == parentHash, "parent block hash somehow wrong");
 
-      bytes32 newroot = Lib_RLPReader.readBytes32(blockNp1[3]);
-      require(assertionRoot != newroot, "asserting that the real state is correct is not a challenge");
+      bytes32 newroot = Lib_RLPReader.readBytes32(decodedHeader[3]);
+      require(assertionRoot != newroot,
+          "asserting that the real state is correct is not a challenge");
+
+      bytes32 txhash    = Lib_RLPReader.readBytes32(decodedHeader[4]);
+      bytes32 coinbase  = bytes32(bytes20(Lib_RLPReader.readAddress(decodedHeader[2])));
+      bytes32 unclehash = Lib_RLPReader.readBytes32(decodedHeader[1]);
+      bytes32 gaslimit  = Lib_RLPReader.readBytes32(decodedHeader[9]);
+      bytes32 time      = Lib_RLPReader.readBytes32(decodedHeader[11]);
 
-      // load starting info into the input oracle
-      // we both agree at the beginning
-      bytes32 txhash = Lib_RLPReader.readBytes32(blockNp1[4]);
-      bytes32 coinbase = bytes32(uint256(Lib_RLPReader.readAddress(blockNp1[2])));
-      bytes32 unclehash = Lib_RLPReader.readBytes32(blockNp1[1]);
-      bytes32 gaslimit = bytes32(Lib_RLPReader.readUint256(blockNp1[9]));
-      bytes32 time = bytes32(Lib_RLPReader.readUint256(blockNp1[11]));
```

Comment:
This change broke the simple demo challenge (but not the fault challenge, interestingly).
(Well at least it's in commit 591cd509ef6494c4cd42bf95f789c816e17b5a76 and I don't see what else it could be.)

Not sure why however. `Lib_RLPReader.readUint256` is defined as `return uint256(readBytes32(_in));`
The address conversation?

Here's how it breaks: the challenge proceeds smoothly until the last step, where the challenger fails as expected, but then the defender *also* fails ("wrong asserted state for defender"), because it uses the exact same state as the challenger (i.e. `assertedState[c.R] == defendedState[c.R]`).

Note that in the simple challenge, the two parties disagree on the very first step — because the challenger uses data from the wrong block, which is only visible after the first step, given that agreement on the initial state is forced by this function.

It shouldn't make sense that they'd end up with the same state hash after the first step though, because the challenger should have computed a wholly different input hash based on the wrong block information, and the difference in input hash should cause the MIPS state hash to be different.

Reply:
Alright, so the problematic diff is:

```
-       bytes32 coinbase = bytes32(uint256(Lib_RLPReader.readAddress(blockNp1[2])));
+       bytes32 coinbase = bytes32(bytes20(Lib_RLPReader.readAddress(blockNp1[2])));
```

The issue is that numbers are little-endian in Solidity, but converting them to bytes reverses the endianness. So we need to zero-extend via a integer conversion before converting to bytes to preserve the number.

>  because it uses the exact same state as the challenger (i.e. assertedState[c.R] == defendedState[c.R]).

As for this, I was just reading the log incorrectly (confusing the log entries for the defender `estimateGas` call as the log entries for the challenger call).